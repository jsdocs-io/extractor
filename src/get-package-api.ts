import { performance } from "node:perf_hooks";
import { join } from "pathe";
import { Bun } from "./bun.ts";
import { getPackageDeclarations } from "./get-package-declarations.ts";
import { getPackageJson } from "./get-package-json.ts";
import { getPackageOverview } from "./get-package-overview.ts";
import { getPackageTypes } from "./get-package-types.ts";
import { getProject } from "./get-project.ts";
import { tempDir } from "./temp-dir.ts";
import type { ExtractedDeclaration } from "./types.ts";

/** `GetPackageApiOptions` contains the options for calling {@link getPackageApi}. */
export interface GetPackageApiOptions {
	/**
  Package to extract the API from.

  This can be either a package name (e.g., `foo`, `@foo/bar`) or
  any other query that can be passed to `bun add` (e.g., `foo@1.0.0`).

  @see {@link https://bun.sh/docs/cli/add | Bun docs}
  */
	pkg: string;

	/**
  Specific subpath to consider in a package.

  If a package has multiple entrypoints listed in the `exports` map property
  of its `package.json`, use `subpath` to select a specific one by its name
  (e.g., `someFeature`).

  @defaultValue `.` (package root)

  @see {@link https://nodejs.org/api/packages.html#subpath-exports | Node.js docs}
  @see {@link https://github.com/lukeed/resolve.exports | resolve.exports docs}
  */
	subpath?: string;

	/**
  Packages can have deeply nested modules and namespaces.

  Use `maxDepth` to limit the depth of the extraction.
  Declarations nested at levels deeper than this value will be ignored.

  @defaultValue 5
  */
	maxDepth?: number;

	/**
	Bun instance used to install the package.

  @defaultValue a new `Bun` instance
  */
	bun?: Bun;
}

/** `PackageApi` contains the data extracted from a package by calling {@link getPackageApi}. */
export interface PackageApi {
	/** Package name (e.g., `foo`, `@foo/bar`). */
	name: string;

	/** Package version number (e.g., `1.0.0`). */
	version: string;

	/**
  Package subpath selected when extracting the API (e.g., `.`, `someFeature`).

  @see {@link ExtractPackageApiOptions.subpath}
  @see {@link https://nodejs.org/api/packages.html#subpath-exports | Node.js docs}
  */
	subpath: string;

	/**
  Type declarations file, resolved from the selected `subpath`,
  that acts as the entrypoint for the package (e.g., `index.d.ts`).
  */
	types?: string;

	/**
  Package description extracted from the `types` file if a
  JSDoc comment with the `@packageDocumentation` tag is found.
  */
	overview?: string;

	/** Declarations exported (or re-exported) by the package. */
	declarations: ExtractedDeclaration[];

	/**
  All packages resolved and installed when installing the package (included).

  @example
  ```ts
  // Installing `foo` brings in also `bar` and `baz` as dependencies.
  ["foo@1.0.0", "bar@2.0.0", "baz@3.0.0"]
  ```
  */
	dependencies: string[];

	/** Timestamp of when the package was analyzed. */
	analyzedAt: string;

	/** Package analysis duration in milliseconds. */
	analyzedIn: number;
}

/**
`getPackageApi` extracts the API from a package.

If the extraction succeeds, `getPackageApi` returns a {@link PackageApi} object.
If the extraction fails, `getPackageApi` throws an error.

Warning: The extraction process is slow and blocks the main thread, using workers is recommended.

@example
```ts
const packageApi = await getPackageApi({
  pkg: "foo",    // Extract API from npm package `foo` [Required]
  subpath: ".",  // Select subpath `.` (root subpath) [Optional]
  maxDepth: 5,   // Maximum depth for analyzing nested namespaces [Optional]
  bun: new Bun() // Bun package manager instance [Optional]
});
console.log(JSON.stringify(packageApi, null, 2));
```

@param options - {@link GetPackageApiOptions}

@returns A {@link PackageApi} object
*/
export async function getPackageApi({
	pkg,
	subpath = ".",
	maxDepth = 5,
	bun = new Bun(),
}: GetPackageApiOptions): Promise<PackageApi> {
	// Normalize options.
	pkg = pkg.trim();
	subpath = subpath.trim() || ".";
	maxDepth = Math.max(1, Math.round(maxDepth));

	// Start performance timer.
	const start = performance.now();

	// Create a temporary directory where to install the package.
	await using dir = await tempDir();
	const cwd = dir.path;

	// Install the package and its direct and third-party dependencies.
	const dependencies = await bun.add(pkg, cwd);

	// Read the package's `package.json`.
	const pkgName = await getInstalledPackageName(cwd);
	const pkgDir = join(cwd, "node_modules", pkgName);
	const pkgJson = await getPackageJson(pkgDir);
	const { name, version } = pkgJson;

	// Find the package's types entry point file, if any.
	const types = getPackageTypes({ pkgJson, subpath });
	if (!types) {
		return {
			name,
			version,
			subpath,
			types: undefined,
			overview: undefined,
			declarations: [],
			dependencies,
			analyzedAt: analyzedAt(),
			analyzedIn: analyzedIn(start),
		};
	}

	// Create TypeScript project.
	const pkgTypes = join(pkgDir, types);
	const { project, indexFile } = getProject({ indexFilePath: pkgTypes, typeRoots: cwd });

	// Get overview.
	const overview = getPackageOverview(indexFile);

	// Extract the declarations exported by the package.
	const declarations = await getPackageDeclarations({ pkgName, project, indexFile, maxDepth });

	// Return the data extracted from the package.
	return {
		name,
		version,
		subpath,
		types,
		overview,
		declarations,
		dependencies,
		analyzedAt: analyzedAt(),
		analyzedIn: analyzedIn(start),
	};
}

async function getInstalledPackageName(cwd: string): Promise<string> {
	// Since `pkg` can contain any argument accepted by Bun's `add` command
	// (e.g., URLs), get the package name from the only dependency listed in
	// the `package.json` file created by Bun in the `cwd` on install.
	const pkgJson = await getPackageJson(cwd);
	return Object.keys(pkgJson.dependencies!).at(0)!;
}

function analyzedAt(): string {
	return new Date().toISOString();
}

function analyzedIn(start: number): number {
	return Math.round(performance.now() - start);
}
