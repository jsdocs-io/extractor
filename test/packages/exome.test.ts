import { expect, test } from "vitest";
import { getPackageApi } from "../../src/get-package-api.ts";

test("exome@2.4.0", async () => {
	expect(
		await getPackageApi({
			pkg: "exome@2.4.0",
		}),
	).toMatchSnapshot({
		analyzedAt: expect.any(String),
		analyzedIn: expect.any(Number),
		dependencies: expect.any(Array),
	});
});

test("exome@2.4.0@ghost", async () => {
	expect(
		await getPackageApi({
			pkg: "exome@2.4.0",
			subpath: "ghost",
		}),
	).toMatchSnapshot({
		analyzedAt: expect.any(String),
		analyzedIn: expect.any(Number),
		dependencies: expect.any(Array),
	});
});
