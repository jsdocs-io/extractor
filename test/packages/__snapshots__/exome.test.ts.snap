// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`exome@2.4.0 1`] = `
{
  "analyzedAt": Any<String>,
  "analyzedIn": Any<Number>,
  "declarations": [
    {
      "constructors": [
        {
          "docs": [],
          "file": "/exome/constructor.d.ts",
          "id": "+class.Exome.constructor",
          "kind": "class-constructor",
          "line": 5,
          "name": "constructor",
          "signature": "constructor();",
        },
      ],
      "docs": [],
      "file": "/exome/constructor.d.ts",
      "id": "+class.Exome",
      "kind": "class",
      "line": 2,
      "methods": [],
      "name": "Exome",
      "properties": [],
      "signature": "class Exome {}",
    },
    {
      "docs": [],
      "file": "/exome/middleware.d.ts",
      "id": "+function.addMiddleware",
      "kind": "function",
      "line": 4,
      "name": "addMiddleware",
      "signature": "addMiddleware: (fn: Middleware) => () => void;",
    },
    {
      "docs": [],
      "file": "/exome/utils/id.d.ts",
      "id": "+function.getExomeId",
      "kind": "function",
      "line": 2,
      "name": "getExomeId",
      "signature": "getExomeId: (store: Exome) => string;",
    },
    {
      "docs": [],
      "file": "/exome/on-action.d.ts",
      "id": "+function.onAction",
      "kind": "function",
      "line": 3,
      "name": "onAction",
      "signature": "onAction: <T extends Exome>(
  Parent: new (...args: any[]) => T,
  action: "NEW" | "LOAD_STATE" | keyof T | null,
  callback: (
    instance: T,
    action: "NEW" | "LOAD_STATE" | keyof T,
    payload: any[],
    error?: Error,
  ) => void,
  type?: "before" | "after",
) => Unsubscribe;",
    },
    {
      "docs": [],
      "file": "/exome/middleware.d.ts",
      "id": "+function.runMiddleware",
      "kind": "function",
      "line": 5,
      "name": "runMiddleware",
      "signature": "runMiddleware: (
  parent: Parameters<Middleware>[0],
  key: Parameters<Middleware>[1],
  args: Parameters<Middleware>[2],
) => (error?: Error) => void;",
    },
    {
      "docs": [],
      "file": "/exome/utils/id.d.ts",
      "id": "+function.setExomeId",
      "kind": "function",
      "line": 3,
      "name": "setExomeId",
      "signature": "setExomeId: (store: Exome, id: string) => void;",
    },
    {
      "docs": [],
      "file": "/exome/subscribe.d.ts",
      "id": "+function.subscribe",
      "kind": "function",
      "line": 3,
      "name": "subscribe",
      "signature": "subscribe: <T extends Exome>(
  store: T,
  fn: (store: T) => void,
) => () => void;",
    },
    {
      "docs": [],
      "file": "/exome/subscribe.d.ts",
      "id": "+function.update",
      "kind": "function",
      "line": 4,
      "name": "update",
      "signature": "update: (store: Exome) => void;",
    },
    {
      "docs": [],
      "file": "/exome/subscribe.d.ts",
      "id": "+function.updateAll",
      "kind": "function",
      "line": 5,
      "name": "updateAll",
      "signature": "updateAll: () => void;",
    },
    {
      "docs": [],
      "file": "/exome/middleware.d.ts",
      "id": "+type.Middleware",
      "kind": "type",
      "line": 2,
      "name": "Middleware",
      "signature": "type Middleware = (
  instance: Exome,
  action: string,
  payload: any[],
) => void | ((error?: Error) => void);",
    },
    {
      "docs": [],
      "file": "/exome/constants.d.ts",
      "id": "+variable.exomeId",
      "kind": "variable",
      "line": 1,
      "name": "exomeId",
      "signature": "const exomeId: Symbol;",
    },
    {
      "docs": [],
      "file": "/exome/constants.d.ts",
      "id": "+variable.exomeName",
      "kind": "variable",
      "line": 2,
      "name": "exomeName",
      "signature": "const exomeName: Symbol;",
    },
  ],
  "dependencies": Any<Array>,
  "name": "exome",
  "overview": undefined,
  "subpath": ".",
  "types": "./exome.d.ts",
  "version": "2.4.0",
}
`;

exports[`exome@2.4.0@ghost 1`] = `
{
  "analyzedAt": Any<String>,
  "analyzedIn": Any<Number>,
  "declarations": [
    {
      "constructors": [],
      "docs": [],
      "file": "/exome/ghost.d.ts",
      "id": "+class.GhostExome",
      "kind": "class",
      "line": 2,
      "methods": [],
      "name": "GhostExome",
      "properties": [],
      "signature": "class GhostExome {}",
    },
  ],
  "dependencies": Any<Array>,
  "name": "exome",
  "overview": undefined,
  "subpath": "ghost",
  "types": "./ghost.d.ts",
  "version": "2.4.0",
}
`;
